src/Symfony/Bundle/FrameworkBundle/Controller/TemplateController.php

```diff
@@ -11,18 +11,39 @@
 
 namespace Symfony\Bundle\FrameworkBundle\Controller;
 
+use Psr\Container\ContainerInterface;
 use Symfony\Component\DependencyInjection\ContainerAwareInterface;
-use Symfony\Component\DependencyInjection\ContainerAwareTrait;
+use Symfony\Component\DependencyInjection\ContainerInterface as SymfonyContainerInterface;
+use Symfony\Component\DependencyInjection\ServiceSubscriberInterface;
 use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Templating\EngineInterface;
+use Twig\Environment;
 
 /**
  * TemplateController.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @final since version 3.4
  */
-class TemplateController implements ContainerAwareInterface
+class TemplateController implements ContainerAwareInterface, ServiceSubscriberInterface
 {
-    use ContainerAwareTrait;
+    protected $container;
+
+    public function __construct(ContainerInterface $container = null)
```

Comment:
Same as for the redirect one. What about passing a Twig and an Engine here instead. By the way, we should probably also deprecate the templating path.